### PR TITLE
Custom set warp permissions without overwriting

### LINKED ISSUE #2863: Add essentials.setwarp.[warpname] and essentials.delwarp.[warpname]
### Feature request

**Feature description**
- Add permission for set especific warps.

- If the player does not have the `essentials.setwarp` or `essentials.setwarp.*` permission and only have the `essentials.setwarp.[warpname]` permission he can only set the warp with name `[warpname]`

- The same goes for deleting a warp.

- The system would be similar to the warps overlay system `essentials.warp.overwrite.*` or `essentials.warp.overwrite.[warpname]`

**How the feature is useful**

- The server admin can give a player permission to set a warp, for example: `essentials.setwarp.PlayerName` then the player can set a warp with his name and announce to his friends.

### PR DESCRIPTION
**Pull request by Haarolean**
Few points:
1. I think there are no changes required for delwarp command suggested in p. 3 of original issue, since granting named permissions should work, unless I missed something.
2. What are the guidelines for i18n? Should I create empty messages for all locales? Or google translate them all?

### CODE REVIEW

File: Essentials/test/com/earth2me/essentials/WarpTest.java
Comment (Haarolean) [inline Essentials/test/com/earth2me/essentials/WarpTest.java:17]: For clarification, I've tested these changes manually as well, with LuckyPerms, if that matters.

Comment (benwoo1110) [reply Essentials/test/com/earth2me/essentials/WarpTest.java:17]: LuckPerms ;)

File: Essentials/src/com/earth2me/essentials/commands/Commanddelwarp.java
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commanddelwarp.java]: Please keep a clean diff and do not change code style or structure in areas of the code that are not immediately relevant to your PR. Code cleanup can be made in a separate PR if you wish.

File: Essentials/src/messages_en.properties
Comment (pop4959) [inline Essentials/src/messages_en.properties]: Don't add translations to files other than the default `messages.properties`, as [Crowdin](https://crowdin.com/project/essentials-2) will overwrite them. If you have translations to submit, it should be made there.

File: Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java:41]: Is this new permission backwards compatible with users on old versions of Essentials prior to this PR? Players with the permission `essentials.setwarp` should be able to create all warps by default. Permission defaults should be defined in `plugin.yml`.

File: Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java
Comment (pop4959) [inline Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java]: Nitpick: You should declare variables just before they get used, and not ahead of time (especially when the function may fail in-between). Helps with readability / understanding the code as well.

File: Essentials/src/messages.properties
Comment (pop4959) [inline Essentials/src/messages.properties]: Maybe it would be better to call this `warpCanCreate` to be explicit, since `warpCreate` could be confused with `warpSet`.

**Review by pop4959 — CHANGES_REQUESTED** [review]
Just a few comments to get you started. I'll review this again more thoroughly sometime later.

File: Essentials/src/messages_en.properties
Comment (Haarolean) [reply Essentials/src/messages_en.properties]: Even better. Reverted.

File: Essentials/src/com/earth2me/essentials/commands/Commanddelwarp.java
Comment (Haarolean) [reply Essentials/src/com/earth2me/essentials/commands/Commanddelwarp.java]: Sorry, auto hard wraps :) Also, the whole class is kinda relevant.

File: Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java
Comment (Haarolean) [reply Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java]: Seems like it can't be null so I placed it near the `warpName`. Whatever, moved lower.

File: Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java
Comment (Haarolean) [reply Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java:41]: Ugh, guess not. Gotta fix

**Comment by ToonTown0909** [comment]
I'll be trying to keep a special eye on this PR, but feel free to reach out if y'all want any changes to `xeya.me`.

File: Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java
Comment (Haarolean) [reply Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java:41]: @pop4959 should we keep `essentials.setwarp` as a fallback permission? I don't see any other ways since it's auto checked by Essentials' command handler. We could grant creating any warps via this permission and allow only specific if user has any *additional* perms like `essentials.warp.set.test`.
`essentials.setwarp` -> allow creating any warp
`essentials.setwarp` + `essentials.warp.set.test` -> allow creating only 'test' warp
Like that?

Comment (pop4959) [reply Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java:41]: How legacy support has been done for other permissions is normally that the base permission node acts as if the user has permission for all sub-nodes, and in this way there is no breakage with existing permissions.

E.g.
`essentials.setwarp=true` -> allow creating any warp
`essentials.setwarp=undefined/false` + `essentials.setwarp.test=true` -> allow creating only `test` warp

Comment (pop4959) [reply Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java:41]: I would like to add that the above is how I would probably try to go about implementing this, however if any other team members want to give another opinion on this that would be appreciated.

Comment (Haarolean) [reply Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java:41]: @pop4959 @md678685 what do you guys think about reusing `per-warp-permission` setting? It's currently used for tp'ing to warps only, we could extent it to warp setting stuff or do we need a new option?

Comment (Octol1ttle) [reply Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java:41]: It would change the existing behavior, and when it's regarding permissions I would like if you kept it unaffected and added a new option instead

Comment (pop4959) [reply Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java:41]: That sounds like a great idea, and if it were just me I would recommend that option. I don't believe minor breaking changes should be a blocking issue when it comes to improving software, but the current norm for Essentials is to make things absolutely idiot proof so that people who don't change their config / permissions *at all* still functions the same after an update. You'll have to wait for @md678685 for a final decision on that since he is the team lead.

Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java:41]: If it weren't for the fact that the behaviour of that config option was well-documented and well-known, I'd agree with pop. However, since this is an entirely separate feature, lumping it in with the existing config option will confuse server owners ("why do I need extra permissions to set warps that I didn't need before?"). I'd recommend putting this under a separate config option (`per-warp-edit-permissions`) that enables per-warp permissions for `setwarp` and `delwarp`.

Comment (ToonTown0909) [reply Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java:41]: I agree with MD that the config options should remain separate.

**Comment by JRoy** [comment]
Closed due to being stale

**Pull request closed without merging**
